mu/modes/web.py

```diff
@@ -0,0 +1,241 @@
+                self.stop_server()
+                return
+            # If needed, save the script.
+            if tab.isModified():
+                self.editor.save_tab_to_file(tab)
+            logger.debug(tab.text())
+            envars = self.editor.envars
+            envars.append(("FLASK_APP", os.path.basename(tab.path)))
+            envars.append(("FLASK_ENV", "development"))
+            args = ["-m", "flask", "run"]
+            cwd = os.path.dirname(tab.path)
+            self.runner = self.view.add_python3_runner("",
+                                                       cwd,
+                                                       interactive=False,
+                                                       envars=envars,
+                                                       python_args=args)
+            logger.debug('Starting Flask app.')
+            self.runner.process.waitForStarted()
+
+    def stop_server(self):
+        """
+        Stop the currently running web server.
+        """
+        logger.debug('Stopping Flask app.')
+        if self.runner:
+            pid = self.runner.process.processId()
+            os.kill(pid, signal.SIGINT)
+            self.runner.process.waitForFinished()
+            self.runner = None
+        self.view.remove_python_runner()
+
+    def stop(self):
+        """
+        Called by logic.quit, so if the server is running, stop it.
+        """
+        self.stop_server()
+
+    def open_file(self, path):
+        """
+        Open the referenced file (html / css).
+        """
+        return read_and_decode(path)
+
+    def load_views(self, event):
+        """
+        Open the directory containing the HTML template views used by Flask.
+
+        This should open the host OS's file system explorer so users can drag
+        new files into the opened folder.
+        """
+        views_dir = os.path.join(self.workspace_dir(), 'templates')
+        logger.info(views_dir)
+        self.editor.load(default_path=views_dir)
+
+    def load_css(self, event):
+        """
+        Open the directory containing the HTML template views used by Flask.
+
+        This should open the host OS's file system explorer so users can drag
+        new files into the opened folder.
+        """
+        css_dir = os.path.join(self.workspace_dir(), 'static', 'css')
+        logger.info(css_dir)
+        self.editor.load(default_path=css_dir)
+
+    def show_images(self, event):
+        """
+        Open the directory containing the static image assets used by Flask.
+
+        This should open the host OS's file system explorer so users can drag
+        new files into the opened folder.
+        """
+        img_dir = os.path.join(self.workspace_dir(), 'static', 'img')
+        logger.info(img_dir)
+        self.view.open_directory_from_os(img_dir)
+
+    def browse(self, event):
+        """
+        Open the default browser to http://127.0.0.1:5080/
+        """
```

Comment:
Should browse warn the user if the server is not running?

Reply:
Good catch. :-)

Reply:
Done.